File: packages/protocol/src/commands.ts
import { z } from 'zod';

import { envelopeSchema, idSchema } from './envelope.js';

/**
 * Commands travel from the UI to the worker. They are deliberately few: the
 * UI can answer an approval, cancel one action, or cancel the conversation.
 * There is no command that grants blanket permissions or changes configuration.
 */
function defineCommand<S extends z.ZodRawShape & { type: z.ZodLiteral<string> }>(shape: S) {
  return envelopeSchema.extend(shape).strict();
}

const approvalRespond = defineCommand({
  type: z.literal('approval.respond'),
  actionId: idSchema,
  approvalId: idSchema,
  decision: z.enum(['approved', 'rejected']),
});

const actionCancel = defineCommand({
  type: z.literal('action.cancel'),
  actionId: idSchema,
});

const conversationCancel = defineCommand({ type: z.literal('conversation.cancel') });

export const commandSchemas = {
  'approval.respond': approvalRespond,
  'action.cancel': actionCancel,
  'conversation.cancel': conversationCancel,
} as const;

export type CommandType = keyof typeof commandSchemas;

export const COMMAND_TYPES = Object.keys(commandSchemas) as readonly CommandType[];

export const commandSchema = z.discriminatedUnion('type', [
  approvalRespond,
  actionCancel,
  conversationCancel,
]);

export type AgentVoiceCommand = z.infer<typeof commandSchema>;

export type CommandOfType<T extends CommandType> = Extract<AgentVoiceCommand, { type: T }>;

export function isCommandType(value: string): value is CommandType {
  return Object.prototype.hasOwnProperty.call(commandSchemas, value);
}

File: apps/worker/src/agent_voice_worker/protocol/commands.py
"""Commands received from the UI on the `agent-voice.commands.v1` topic.

Commands travel from the UI to the worker. They are deliberately few: the
UI can answer an approval, cancel one action, or cancel the conversation.
There is no command that grants blanket permissions or changes
configuration. Mirrors `packages/protocol/src/commands.ts`.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import ConfigDict, Field

from agent_voice_worker.protocol.envelope import Envelope, IdStr


class _Command(Envelope):
    model_config = ConfigDict(extra="forbid")


class ApprovalRespond(_Command):
    type: Literal["approval.respond"]
    actionId: IdStr
    approvalId: IdStr
    decision: Literal["approved", "rejected"]


class ActionCancel(_Command):
    type: Literal["action.cancel"]
    actionId: IdStr


class ConversationCancel(_Command):
    type: Literal["conversation.cancel"]


AgentVoiceCommand = Annotated[
    ApprovalRespond | ActionCancel | ConversationCancel, Field(discriminator="type")
]

COMMAND_MODELS: dict[str, type[_Command]] = {
    "approval.respond": ApprovalRespond,
    "action.cancel": ActionCancel,
    "conversation.cancel": ConversationCancel,
}

COMMAND_TYPES: tuple[str, ...] = tuple(COMMAND_MODELS.keys())


def is_command_type(value: str) -> bool:
    return value in COMMAND_MODELS


def command_model_for(command_type: str) -> type[_Command]:
    return COMMAND_MODELS[command_type]
